ta/pkcs11/src/processing_symm.c

```diff
@@ -183,27 +222,69 @@ allocate_tee_operation(struct pkcs11_session *session,
 	return tee2pkcs_error(res);
 }
 
+static enum pkcs11_rc pkcs2tee_load_hashed_attr(TEE_Attribute *tee_ref,
+						uint32_t tee_id,
+						struct pkcs11_object *obj,
+						enum pkcs11_attr_id pkcs11_id,
+						uint32_t tee_algo,
+						void *hash_ptr,
+						uint32_t *hash_size)
+{
+	TEE_OperationHandle handle = TEE_HANDLE_NULL;
+	void *a_ptr = NULL;
+	uint32_t a_size = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (get_attribute_ptr(obj->attributes, pkcs11_id, &a_ptr, &a_size))
+		return PKCS11_CKR_FUNCTION_FAILED;
+
+	/* Calculate hash of the attribute */
+	res = TEE_AllocateOperation(&handle, tee_algo, TEE_MODE_DIGEST, 0);
+	if (res) {
+		EMSG("TEE_AllocateOp. failed %#"PRIx32"", tee_algo);
+		return PKCS11_CKR_FUNCTION_FAILED;
+	}
+
+	res = TEE_DigestDoFinal(handle, a_ptr, a_size, hash_ptr, hash_size);
+	if (res) {
+		EMSG("TEE_DigestOp. failed %#"PRIx32"", tee_algo);
+		TEE_FreeOperation(handle);
+		return PKCS11_CKR_FUNCTION_FAILED;
+	}
+
+	TEE_InitRefAttribute(tee_ref, tee_id, hash_ptr, *hash_size);
+
+	TEE_FreeOperation(handle);
+	return PKCS11_CKR_OK;
+}
+
 static enum pkcs11_rc load_tee_key(struct pkcs11_session *session,
 				   struct pkcs11_object *obj,
 				   struct pkcs11_attribute_head *proc_params)
 {
+	uint32_t algo = 0;
 	TEE_Attribute tee_attr = { };
 	size_t object_size = 0;
+	uint32_t key_size = 0;
 	uint32_t tee_key_type = 0;
 	enum pkcs11_key_type key_type = 0;
 	enum pkcs11_rc rc = PKCS11_CKR_OK;
 	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t max_key_size = 0;
+	uint32_t min_key_size = 0;
+	void *hash_ptr = NULL;
+	uint32_t hash_size = 0;
 
 	if (obj->key_handle != TEE_HANDLE_NULL) {
 		/* Key was already loaded and fits current need */
 		goto key_ready;
 	}
 
-	if (!pkcs2tee_load_attr(&tee_attr, TEE_ATTR_SECRET_VALUE,
-				obj, PKCS11_CKA_VALUE)) {
-		EMSG("No secret found");
-		return PKCS11_CKR_FUNCTION_FAILED;
-	}
+	object_size = get_object_key_bit_size(obj);
+	if (!object_size)
+		return PKCS11_CKR_GENERAL_ERROR;
+
+	key_size = object_size / 8;
```

Comment:
`key_size` is only used once the switch/case. I suggest line 315 uses `object_size / 8` straight.

Reply:
Changed